# HW 1

## 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [1]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv"
df = pd.read_csv(url)

# Check for missing values in each column
missing_values = df.isnull().sum()

# Display columns with missing values
print(missing_values)

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64


### Summary of Interaction:

In this interaction, I sought assistance in obtaining a CSV dataset with missing values for use in a data analysis task. Initially, I asked for the Bigfoot Sightings dataset, which I believed contained missing data. However, the links provided for that dataset were not functional. The assistant then provided alternative datasets, and we eventually used the California Housing Prices dataset from a reliable GitHub source.

After obtaining the dataset, I asked for help in verifying the presence of missing values. The assistant provided a Python code snippet using pandas, which checks for missing values in each column of the dataset. The "total_bedrooms" column in the California Housing dataset contains missing values, which makes this dataset suitable for my analysis.

The assistant's guidance allowed me to successfully load the dataset and confirm that it has missing values, which fulfills the requirements of my assignment.

https://chatgpt.com/share/f0f5c4df-0cbc-4b14-9b50-67ce170442e6

## 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then

    1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,
    2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

In [2]:
# Get the shape of the DataFrame
num_rows, num_columns = df.shape

# Print the number of rows and columns
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

Number of rows: 20640
Number of columns: 10


**Observations** - The rows in a data set, typically the objects in the data set.

**Variables** - The columns in the data set, being the attributes of the different observations.

### Summary of Interaction:

In this session, the user sought assistance with working on a dataset in Python using the pandas library. The dataset in question was accessed via the URL: "https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv". The user requested guidance on how to determine and print the number of rows and columns in the dataset.

To address this request, the following steps were provided:

    Import the pandas library.
    Read the dataset from the provided URL using pd.read_csv().
    Use the shape attribute of the DataFrame to obtain the number of rows and columns.

Sample code was provided:

In [3]:
import pandas as pd

# Read the dataset from the URL
url = "https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv"
df = pd.read_csv(url)

# Get the shape of the DataFrame
num_rows, num_columns = df.shape

# Print the number of rows and columns
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

Number of rows: 20640
Number of columns: 10


Additionally, the user inquired about the definitions of "observations" and "variables" in the context of their dataset. It was explained that:

    Observations refer to the individual rows in the dataset, each representing a single instance or unit of analysis (e.g., a single house in a housing dataset).
    Variables refer to the columns in the dataset, each representing a characteristic or attribute of the observations (e.g., number of bedrooms, size of the house).
    
https://chatgpt.com/share/a642f164-6603-419b-a507-e5bbbd1d5bdc 

## 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [4]:
# Display the first few rows of the dataset
print("First 5 rows of the dataset:")
print(df.head())

# Get a summary of the dataset (data types and non-null counts)
print("\nInfo about the dataset:")
print(df.info())

# Get statistical summary of the dataset (for numerical columns)
print("\nStatistical summary of the dataset:")
print(df.describe())

# Show unique value counts for each column
print("\nValue counts for each column:")
for column in df.columns:
    print(f"\nColumn: {column}")
    print(df[column].value_counts().head(10))  # Display top 10 most frequent values

First 5 rows of the dataset:
   longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY  

Info about the dataset:
<class 'pandas.core.f

### Summary of Interaction:

In this discussion, I explored methods to summarize a dataset using Python and the pandas library. I used the housing dataset available at the link https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv.

To provide a basic summary of the dataset, I employed functions like:

    df.head() to preview the first few rows.
    df.info() to obtain information about the dataset's columns, such as data types and non-null counts.
    df.describe() to generate a statistical summary for numerical columns, including metrics like mean, standard deviation, and percentiles.
    df.isnull().sum() to check for missing values.

Additionally, I used df['column'].value_counts() to display the frequency of unique values in each column, which is especially useful for categorical data. A loop was used to provide value counts for each column, displaying the top 10 most frequent values.

https://chatgpt.com/share/8db02f12-252e-4cb9-9d9c-730cc1cd7536

## 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

In [5]:
# Check the shape of the dataset
print("\nShape:")
print(df.shape)  # e.g., (20640, 10) -> 20640 rows, 10 columns

# Describe the dataset
print("\nDescription:")
print(df.describe())


Shape:
(20640, 10)

Description:
          longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000     

##### a) The number of columns analyzed:
Non-numeric values are not included in the "df.describe()" method as it only deals with numeric values. This means that despite df.shape stating that the data set has 10 columns, the df.describe() method only lists 9, as the "Ocean Proximity" column in the data set holds a non-numeric data type, and is therefore excluded.
##### b) the values it reports in the "count" column
Though most of this data set does not contain missing values, the "total bedrooms" column does have some missing values. This causes a discrepancy between the number of rows stated by the df.shape attribute and the count of values in the total bedrooms column stated by the df.describe() method.

### Summary of Interaction

In this conversation, I learned how to explain the discrepancies between the output of df.shape and df.describe() when analyzing a dataset in Python using Pandas. The dataset in question contains both numeric and non-numeric variables, as well as missing values in numeric columns.

    Non-numeric Columns: The function df.describe() by default only analyzes numeric columns. Non-numeric columns (such as strings or categorical variables) are excluded from the summary statistics, which explains why fewer columns are reported compared to df.shape, which includes all columns.

    Missing Values: The "count" column in df.describe() shows the number of non-missing (non-NaN) values for each column. If a column has missing values, the count will be lower than the total number of rows (as given by df.shape). This can cause the reported count in df.describe() to be less than the total number of rows in the dataset.

By combining these two explanations, I can account for any differences between the overall structure of the dataset and the summary statistics provided by Pandas' df.describe() function.

https://chatgpt.com/share/f985a6e6-bf17-408f-a85d-02b90e66dae1

## 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

    - an "attribute", such as df.shape which does not end with ()
    - and a "method", such as df.describe() which does end with ()

**Attributes** - A property/characteristic of an object. Comparable to a variable, however an attribute is "owned" or "contained" within an object.

**Method** - A function that is native to an object and that executes an action on the object.

**Key Differences**

- Both are "owned" by an object, however attributes are like variables, while methods are like functions.

- An attribute does not require parentheses to be accessed because it does not have parameters (ex: df.shape). Oppositely, methods are functions that require inputs (or its default parameter), so it must be used with parentheses (ex: df.describe()).
    
### Summary of Interaction

In this conversation, I explored the difference between attributes and methods in Python, particularly when using the pandas library:

    Attributes (e.g., df.shape) represent properties of an object and store information about it. They are accessed without parentheses. For instance, df.shape returns the dimensions (rows and columns) of a DataFrame.

    Methods (e.g., df.describe()) are functions associated with an object that perform actions or computations. They are called with parentheses. For example, df.describe() computes and returns summary statistics for numerical columns in the DataFrame.

The key distinction is that attributes provide stored information directly, while methods execute code to perform tasks and return results.

https://chatgpt.com/share/44d177d6-0654-4b68-b754-544da5bac5a5

## 6. The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

**Count** - The number of non-null (non-missing) observations in each column.

**Mean** - The arithmetic average of the values in each column. It is calculated as the sum of all the values divided by the count of the values.

**Standard Deviation** (std) - A measure of the dispersion or spread of the data. It tells how much the values deviate from the mean. The larger the standard deviation, the more spread out the data is.

**Minimum (min)** - The smallest value in each column.

**25th Percentile (25%)** - Also known as the first quartile (Q1). This value represents the point below which 25% of the data falls. It gives an idea of the lower bound of the middle 50% of the data.

**50th Percentile (50%)** - Also known as the median. This is the middle value of the data set, where 50% of the values are below it and 50% are above it.

**75th Percentile (75%)** - Also known as the third quartile (Q3). This value represents the point below which 75% of the data falls. It gives an idea of the upper bound of the middle 50% of the data.

**Maximum (max)** - The largest value in each column.

### Summary of Interaction

In this conversation, the summary statistics generated by the df.describe() method in Python's pandas library were discussed. The method provides the following key metrics:

    Count: Number of non-null observations.
    Mean: The average value of the data.
    Standard Deviation (std): A measure of data spread from the mean.
    Minimum (min): The smallest value in the dataset.
    25th Percentile (25%): The value below which 25% of the data falls (first quartile).
    50th Percentile (50%): The median, or the middle value of the data.
    75th Percentile (75%): The value below which 75% of the data falls (third quartile).
    Maximum (max): The largest value in the dataset.

These statistics provide insights into the distribution, central tendency, and variability of the dataset being analyzed.

https://chatgpt.com/share/c50bd834-fed3-41f5-b5d7-992ed779661a

## 7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

    1. Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']
    2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()
    3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important
    4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

**1. Use case favoring df.dropna()**

df.dropna() automatically deletes any row (0/default) or column (1) that contains missing data (NaN). Therefore, it is favourable when trying to clean a data set of missing values, especially in the case that there are multiple rows/columns that contain missing values

**2. Use case favoring del df['col']**

del df['col'] deletes the named column in the square brackets. This is a favourable operation if you are trying to delete a single and specific column for any reason, especially those other than the presence of missing values.

**3. Applying del df['col'] before df.dropna()**

Applying del df['col'] before df.dropna() when both are used together could be important in the scenario where you are intending on deleting rows with missing values as well as a specific column. If there is a row that is missing a value in the column you intend to drop but not in any other, using df.dropna() first would delete that row for no reason, as it would not have a missing value had del df['col'] been used first. By using del df['col'] before df.dropna() you can avoid this situation and retain that row of data that would have been erased. 

**4. Remove all missing data from the "Housing Dataset"**

As learned in previous questions, the only missing data that the data set contains is in the "total_bedrooms" column. This column is missing 207 values. Here, I am faced with the decision between using df.dropna() to delete all of the rows that contain a missing value in this column, or to use del df['col'] and delete the column altogether. My intention is to delete the least amount of data as possible, which led me to decide to use df.dropna() to delete the 207 rows that are missing a value. The justification behind this is that by deleting the 207 rows, I am only deleting 2,070 values (207 rows x 10 columns). On the other hand, by using  del df['col'] to erase the "total_bedrooms" column would erase 20640 values (20640 rows x 1 column).

In [6]:
# Display the before the deletion
print("\nBefore:\n")

# Check for missing values in each column
missing_values = df.isnull().sum()

# Display columns with missing values
print(missing_values)

# Get the shape of the DataFrame
num_rows, num_columns = df.shape

# Print the number of rows and columns
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

newdf = df.dropna()

print("\nAfter:\n")

# Check for missing values in each column
new_missing_values = newdf.isnull().sum()

# Display columns with missing values
print(new_missing_values)

# Get the shape of the DataFrame
new_num_rows, new_num_columns = newdf.shape

# Print the number of rows and columns
print(f"Number of rows: {new_num_rows}")
print(f"Number of columns: {new_num_columns}")


Before:

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64
Number of rows: 20640
Number of columns: 10

After:

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64
Number of rows: 20433
Number of columns: 10


### Summary of Interaction

In this conversation, we explored the difference between del df['col'] and df.dropna() in pandas and why the order of applying them matters.

    del df['col'] removes a specific column from the DataFrame permanently.
    df.dropna() removes rows or columns containing missing values (NaN).

Importance of Order:

Applying del df['col'] before df.dropna() is crucial when the column to be deleted contains many missing values. Deleting the column first prevents unnecessary row loss during dropna(), ensuring only relevant data is considered. This approach also improves efficiency and ensures valid rows aren't dropped due to irrelevant columns.

https://chatgpt.com/share/d29dbd48-a5d5-46e3-9b0b-96451fbc794d

#### Note: I wrote the code here without the chatbot using past pieces of code and my own understanding. 

## 8. Give brief explanations in your own words for any requested answers to the questions below

1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

2. Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT

    - Forget to include import pandas as pd in your code. Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error. When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is "NameError: name 'pd' is not defined"

    - Mistype "titanic.csv" as "titanics.csv"
        If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv' (assuming the file is indeed not present). Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces

    - Try to use a dataframe before it's been assigned into the variable
        You can simulate this by just misnaming the variable. For example, if you should write df.groupby("col1")["col2"].describe() based on how you loaded the data, then instead write DF.groupby("col1")["col2"].describe(). Make sure you've fixed your file name so that's not the error any more

    - Forget one of the parentheses somewhere the code
        For example, if the code should be pd.read_csv(url) the change it to pd.read_csv(url

     - Mistype one of the names of the chained functions with the code. For example, try something like df.group_by("col1")["col2"].describe() and df.groupby("col1")["col2"].describle()

     - Use a column name that's not in your data for the groupby and column selection. For example, try capitalizing the columns for example replacing "sex" with "Sex" in titanic_df.groupby("sex")["age"].describe(), and then instead introducing the same error of "age"

      - Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question. For example, something like titanic_df.groupby(sex)["age"].describe(), and then titanic_df.groupby("sex")[age].describe()



#### 1.

"df.groupby("col1")["col2"].describe()" organizes the data set based on the common values in a specified column (that in the round parentheses). It then provides same statistics as the ".describe()" method for the column specified in the square brackets, each row being the different group that were previously created. 

For example, if two columns in a data set are the animal's species and the animal's weight, applying "df.groupby("species")["weight"].describe()" would first organize all of the animals by their species (ex: cow, pig, sheep), and then provide the .describe() statistics for the weight column of each group's animals.

In the examples of my previously used "housing" data set and the "titanic" data set:

In [3]:
import pandas as pd

# Load the housing dataset
url1 = "https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv"
df1 = pd.read_csv(url1)

# Group data by proximity to the ocean and analyze the total number of rooms for each group
print(df1.groupby("ocean_proximity")["total_rooms"].describe())

# Load the titanic dataset
url2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df2 = pd.read_csv(url2)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df2.groupby("survived")["age"].describe()

                  count         mean          std    min      25%     50%  \
ocean_proximity                                                             
<1H OCEAN        9136.0  2628.343586  2160.463696   11.0  1464.00  2108.0   
INLAND           6551.0  2717.742787  2385.831111    2.0  1404.00  2131.0   
ISLAND              5.0  1574.600000   707.545264  716.0   996.00  1675.0   
NEAR BAY         2290.0  2493.589520  1830.817022    8.0  1431.25  2083.0   
NEAR OCEAN       2658.0  2583.700903  1990.724760   15.0  1505.00  2195.0   

                     75%      max  
ocean_proximity                    
<1H OCEAN        3141.00  37937.0  
INLAND           3216.00  39320.0  
ISLAND           2127.00   2359.0  
NEAR BAY         3029.75  18634.0  
NEAR OCEAN       3109.00  30405.0  


count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

### Summary of Interaction

Summary of Titanic Dataset Analysis

In this interaction, we explored using pandas' groupby and describe functions for analyzing the Titanic dataset.

1. df.groupby("col1")["col2"].describe():
        Groups data by col1 (e.g., Pclass), then computes summary statistics (count, mean, std, min, max, percentiles) for col2 (e.g., Fare) within each group.

    Example: df.groupby("Pclass")["Fare"].describe() shows fare statistics for each passenger class, revealing differences in fare distributions.

2. Comparison with df.describe():
        df.describe() provides summary statistics for all numerical columns across the entire dataset, without grouping.
        df.groupby("Pclass")["Fare"].describe() focuses on a specific column (Fare) and groups by Pclass, offering more detailed insights by class.

This analysis highlights how groupby allows for more granular, group-specific insights compared to the overall summary from df.describe().

https://chatgpt.com/share/eb7252da-e992-4c5e-9292-1d7672ab7c4a



#### 2.

The values in the "count" column that result from executing "df.groupby("col1")["col2"].describe()" provide the number of values that fall under each category that the data has been grouped into. 

The count column in the description of both the overall data and the specific groups has the function of determining the number of non-missing values. The difference is that counting the values in a sorted data set has the advantage of providing the  distribution of values for of the created groups. 

In the above example, "df.groupby("ocean_proximity")["total_rooms"].describe()" was used on the housing data set. By doing this, the number of values in the "total_rooms" can be seen, however the number of values is additionally distributed into each "ocean_proximity" category. 

#### 3.

#### NOTE: The kernel must be restarted to display each of the following sample errors:

In [2]:
# Error 1 - Not loading the pandas library

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupby("survived")["age"].describe()

NameError: name 'pd' is not defined

In [3]:
# Error 2 - Typo in the link

import pandas as pd

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupby("survived")["age"].describe()

HTTPError: HTTP Error 404: Not Found

In [2]:
# Error 3 - Data frame of undefined variable

import pandas as pd

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
DF.groupby("survived")["age"].describe()

NameError: name 'DF' is not defined

In [3]:
# Error 4 - Forgetting parentheses

import pandas as pd

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupby"survived")["age"].describe()

SyntaxError: unmatched ')' (3161805177.py, line 10)

In [4]:
# Error 5 - Mispelling a function

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupbuy("survived")["age"].describe()

AttributeError: 'DataFrame' object has no attribute 'groupbuy'

In [5]:
# Error 6 - Mispelling column names

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupby("surviyed")["age"].describe()

KeyError: 'surviyed'

In [6]:
# Error 6 - Mispelling column names

# Load the titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group data by whether or not the passenger survived and analyze the total ages of each group
df.groupby(survived)["age"].describe()

NameError: name 'survived' is not defined

### Observations:

For all provided errors, ChatGPT was able to immediately provide the correct solution to the issue, along with the completed and correct code. This was extremely effective and useful for trouble shooting the errors in the code.

Regarding searching for solutions on Google, it was possible to find help for the different errors, however ChatGPT was undeniably more effective and simple to use. One primary advantage of ChatGPT is that it corrected the exact code for my project, as opposed to Google informing me of the general formats that the code should have (after which I would need to apply my data to that format). 

### Summary of Interactions:

In this conversation, the user encountered several errors while working with a dataset from the Titanic. The issues and their resolutions included:

    NameError for pd: The user initially received a NameError because the pandas library was not imported with the alias pd. The solution was to add import pandas as pd at the beginning of the script.

    HTTPError 404: The user faced a 404 Not Found error due to an incorrect URL for the dataset. The correct URL was provided, which resolved the issue.

    Syntax Error in groupby: A SyntaxError occurred because of a missing parenthesis in the groupby method. This was fixed by adding the missing parenthesis.

    AttributeError for groupbuy: The user mistyped groupbuy instead of groupby. Correcting the typo resolved the issue.

    KeyError for Column Name: A KeyError was raised due to a typo in the column name "surviyed" instead of "survived". The correct column name was used to fix the error.

    NameError for Unquoted Column Name: Finally, the user received a NameError because the column name was used without quotes. The correct usage included quotes around the column name "survived".

Each issue was addressed by correcting typos or ensuring proper syntax and usage of Python and pandas methods.

https://chatgpt.com/share/66e3b6e0-231c-8012-a5ff-41c9712b86ca

## 9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Yes. ChatGPT and the textbook have helped me to understand the code and concepts we have learned.
